src/idl_gen_fbs.cpp

```diff
@@ -39,6 +43,97 @@ static std::string GenType(const Type &type, bool underlying = false) {
   }
 }
 
+static std::unordered_map<std::string, voffset_t> MapStructId(
+    const StructDef &struct_def, IDLOptions::ProtoIdGapAction gap_action) {
+  using FieldIdNamePair = std::pair<std::string, std::string>;
+  std::vector<FieldIdNamePair> ids;
+
+  bool field_with_id = false;
+
+  for (const auto &field : struct_def.fields.vec) {
+    if (field->attributes.Lookup("id")) {
+      field_with_id = true;
+      if (field->attributes.Lookup("id")->constant.empty() &&
+          field->value.type.base_type != BASE_TYPE_UNION)
+        ids.push_back(std::make_pair("-1", field->name));
+      else {
+        ids.push_back(std::make_pair(field->attributes.Lookup("id")->constant,
+                                     field->name));
+
+        // Check for non positive id number
+        if (!field->attributes.Lookup("id")->constant.empty()) {
+          voffset_t id = 0;
+          bool done = StringToNumber(
+              field->attributes.Lookup("id")->constant.c_str(), &id);
+          if (!done)
+            throw std::runtime_error("Field id in struct" + struct_def.name +
+                                     "has a non positive number value\n");
+        }
+      }
+    } else {
+      ids.push_back(std::make_pair(std::string(), field->name));
+    }
+  }
+
+  // None of the fields has id
+  if (!field_with_id) return {};
+
+  // Sort ids, consider 0 if the field does not have an id.
+  std::sort(ids.begin(), ids.end(),
+            [](const FieldIdNamePair &lhs, const FieldIdNamePair &rhs) {
+              auto a_id = lhs.first.empty() ? 0 : std::stoi(lhs.first);
+              auto b_id = rhs.first.empty() ? 0 : std::stoi(rhs.first);
+              return a_id < b_id;
+            });
+
+  for (auto it = std::next(ids.begin()); it != ids.end(); it++) {
+    if (std::find(struct_def.reserved_ids.begin(),
+                  struct_def.reserved_ids.end(),
+                  stoi(it->first)) != struct_def.reserved_ids.end())
+      throw std::runtime_error("Field " + it->second + " with id " + it->first +
+                               " in struct " + struct_def.name +
+                               " uses id from reserved ids\n");
+
+    // Check for twice use of ids
+    if (!it->first.empty() && std::prev(it)->first == it->first)
+      throw std::runtime_error("Fields " + it->second + " and " +
+                               std::prev(it)->second + " with use id " +
+                               it->first + " in struct" + struct_def.name +
+                               " twice\n");
+
+    // Check for gap between ids
```

Comment:
I know its convenient to check everything in one loop, but it suffers from poor readability. Since this isn't performance-related code,  and the input is bounded to a reason max field id, I think separate these checks out to their own static function that do one thing on the preprocessed data would be better.